libsolidity/ast/ASTBoogieUtils.cpp

```diff
@@ -1282,51 +1270,205 @@ void ASTBoogieUtils::deepCopyStruct(StructDefinition const* structDef,
 	}
 }
 
-ASTBoogieUtils::FreezeResult ASTBoogieUtils::freeze(bg::Expr::Ref bgExpr,
-		Expression const* expr, ASTNode const* assocNode, BoogieContext& context)
+ASTBoogieUtils::PackResult ASTBoogieUtils::packToLocalPtr(Expression const* expr, bg::Expr::Ref bgExpr, BoogieContext& context)
 {
-	if (auto varExpr = dynamic_pointer_cast<bg::VarExpr const>(bgExpr))
+	PackResult result {nullptr, {}};
+	packInternal(expr, bgExpr, context, result);
+	if (result.ptr)
+		return result;
+
+	context.reportError(expr, "Unsupported expression for packing into local pointer");
+	return PackResult{context.tmpVar(bg::Decl::typee(ERR_TYPE)), {}};
+}
+
+void ASTBoogieUtils::packInternal(Expression const* expr, bg::Expr::Ref bgExpr, BoogieContext& context, PackResult& result)
+{
+	// Function calls return pointers, no need to pack, just copy the return value
+	if (dynamic_cast<FunctionCall const*>(expr))
 	{
-		return FreezeResult{bgExpr, {}, {}};
+		auto ptr = context.tmpVar(ASTBoogieUtils::mappingType(context.intType(256), context.intType(256)));
+		result.ptr = ptr;
+		result.stmts.push_back(bg::Stmt::assign(ptr->getRefTo(), bgExpr));
+		return;
```

Comment:
`result.ptr` is a declaration, not an expression. We could modify it to be an expression so that we don't need this extra declaration and assignment if no actual packing is required. However, then we would have to add the declaration as well in the structure, and at every place where we call `pack`, we would have to handle the declaration and the expression as well. My decision was to keep usage of `pack` simpler with the tradeoff that in this one case an unnecessary new declaration is created.